phantazm-zombies/src/main/java/com/github/phantazmnetwork/zombies/equipment/gun/shoot/fire/HitScanFirer.java

```diff
@@ -89,13 +115,23 @@ public record Data(@NotNull Key endSelectorKey, @NotNull Key targetFinderKey,
 
     private final Collection<ShotHandler> shotHandlers;
 
+    /**
+     * Creates a {@link HitScanFirer}.
+     * @param entitySupplier A {@link Supplier} for the {@link Entity} shooter
+     * @param endSelector The {@link HitScanFirer}'s {@link ShotEndpointSelector}
+     * @param targetFinder The {@link HitScanFirer}'s {@link TargetFinder}
+     * @param shotHandlers The {@link HitScanFirer}'s {@link ShotHandler}s
+     */
     public HitScanFirer(@NotNull Supplier<Optional<? extends Entity>> entitySupplier,
                         @NotNull ShotEndpointSelector endSelector, @NotNull TargetFinder targetFinder,
                         @NotNull Collection<ShotHandler> shotHandlers) {
         this.entitySupplier = Objects.requireNonNull(entitySupplier, "entitySupplier");
         this.endSelector = Objects.requireNonNull(endSelector, "endSelector");
         this.targetFinder = Objects.requireNonNull(targetFinder, "targetFinder");
         this.shotHandlers = Objects.requireNonNull(shotHandlers, "shotHandlers");
+        for (ShotHandler shotHandler : shotHandlers) {
+            Objects.requireNonNull(shotHandler, "shotHandlers shotHandler");
```

Comment:
null objects can be added anyway after this check (unless you make a defensive copy) so i don't really see the point of doing this

Reply:
I do this in a few places, should I do defensive copies or just get rid of collection verification entirely

Reply:
(note: List.copyOf checks for nulls and produces a defensive copy so if anything you should probably be using that)

Reply:
that being said, i personally prefer no collection verification as it just seems like extra effort unless you're already doing a defensive copy

Reply:
I'll do defensive copies where applicable but for records I'd have to change them to classes, or do I just leave them be

Reply:
no point changing the records to classes, that's cringe